stl/inc/syncstream

```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _SYNCSTREAM_
+#define _SYNCSTREAM_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifndef __cpp_lib_syncbuf
+#pragma message("The contents of <syncstream> are available only with C++20 syncbuf support.")
+#else // ^^^ !defined(__cpp_lib_syncbuf) / defined(__cpp_lib_syncbuf) vvv
+#include <map>
+#include <memory>
+#include <shared_mutex>
+#include <streambuf>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#pragma push_macro("emit")
+#undef new
+#undef emit
+
+_STD_BEGIN
+
+extern "C" shared_mutex& __cdecl _Get_mutex_for_instance(void* _Ptr);
+extern "C" void __cdecl _Add_mutex_for_instance_or_increment(void* _Ptr) noexcept;
+extern "C" void __cdecl _Delete_mutex_for_instance_or_decrement(void* _Ptr) noexcept;
+
+// CLASS TEMPLATE basic_syncbuf
+template <class _Elem, class _Traits, class _Alloc>
+class basic_syncbuf : public basic_streambuf<_Elem, _Traits> {
+public:
+    using int_type       = typename _Traits::int_type;
+    using pos_type       = typename _Traits::pos_type;
+    using off_type       = typename _Traits::off_type;
+    using allocator_type = _Alloc;
+    using streambuf_type = basic_streambuf<_Elem, _Traits>;
+
+    using pointer   = typename allocator_traits<_Alloc>::pointer;
+    using size_type = typename allocator_traits<_Alloc>::size_type;
+
+    basic_syncbuf() = default;
+
+    explicit basic_syncbuf(streambuf_type* _Strbuf) : basic_syncbuf(_Strbuf, _Alloc{}) {}
+
+    basic_syncbuf(streambuf_type* _Strbuf, const _Alloc& _Al_) : _Wrapped(_Strbuf), _Al(_Al_) {
+        if (_Wrapped != nullptr) {
+            _Add_mutex_for_instance_or_increment(_Wrapped);
+        }
+        _Init();
+    }
+
+    basic_syncbuf(basic_syncbuf&& _Right) {
+        _Al = _STD move(_Right._Al);
+        _Assign_rv(_STD move(_Right));
+    }
+
+    ~basic_syncbuf() {
+        _Emit();
+        _Tidy();
+    }
+
+    basic_syncbuf& operator=(basic_syncbuf&& _Right) {
+        _Emit();
+        _Move_assign(_STD move(_Right), _Choose_pocma<_Alloc>{});
+        return *this;
+    }
+
+    void swap(basic_syncbuf& _Right) noexcept {
+        if (this != _STD addressof(_Right)) {
+            _Pocs(_Getal(), _Right._Getal());
+            _Swap_except_al(_Right);
+        }
+    }
+
+    bool emit() {
+        if (_Wrapped == nullptr) {
+            return false;
+        }
+
+        bool _Result                 = true;
+        const size_type _Data_size   = _Get_data_size();
+        const pointer _Begin_seq_ptr = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        if (_Data_size > 0 || _Record_sync) {
+            scoped_lock _Guard(_Get_mutex_for_instance(_Wrapped));
+
+            if (_Data_size > 0 && _Data_size != _Wrapped->sputn(_Begin_seq_ptr, _Data_size)) {

```

Comment:
This line produces:
```C++
 warning C4389: '!=': signed/unsigned mismatch
```
This made me thinking should I keep `size_type` as `allocator_traits<_Alloc>::size_type` or change it to `streamsize` since `streamsize` could technically be larger than `size_t`

Reply:
Not just technically - our `streamsize` is always `long long` even on 32-bit platforms:
https://github.com/microsoft/STL/blob/ac4fde764ae716512bdd67e0e9577bd53947bc3d/stl/inc/iosfwd#L49

Reply:
It's obvious then, I'll do the necessary changes to use `streamsize` as the `size_type` of `basic_syncbuf`.

Reply:
After re-consideration, using `size_type` is more plausible since the data held by the `streambuf` can not exceed `max_size` of the allocator anyway. a `static_cast` of the returned value from `sputn` should be enough.